discovery_engine_core/core/src/engine.rs

```diff
@@ -1065,230 +1114,290 @@ enum SearchBy<'a> {
 }
 
 #[cfg(test)]
-mod tests {
-    use crate::{
-        ranker,
-        stack::{ops::MockOps, Data},
-    };
+pub(crate) mod tests {
+    use std::mem::size_of;
 
-    use crate::document::tests::mock_generic_article;
-    use std::{mem::size_of, sync::Arc};
+    use async_once_cell::OnceCell;
+    use tokio::sync::{MappedMutexGuard, Mutex, MutexGuard};
     use wiremock::{
         matchers::{method, path},
         Mock,
         MockServer,
         ResponseTemplate,
     };
 
+    use crate::{document::tests::mock_generic_article, stack::ops::MockOps};
+
     use super::*;
 
     #[test]
     fn test_usize_not_to_small() {
         assert!(size_of::<usize>() >= size_of::<u32>());
     }
 
-    #[tokio::test]
-    async fn test_cross_stack_deduplication() {
-        let mock_server = MockServer::start().await;
-        let tmpl = ResponseTemplate::new(200)
-            .set_body_string(include_str!("../test-fixtures/newscatcher/duplicates.json"));
-
-        Mock::given(method("GET"))
-            .and(path("/newscatcher/headlines-endpoint-name"))
-            .respond_with(tmpl)
-            .mount(&mock_server)
-            .await;
+    /// A shared test engine.
+    ///
+    /// The systems are only initialized once to save time and the states can be configured for each
+    /// test via [`init_engine()`].
+    static ENGINE: OnceCell<Mutex<(MockServer, Engine)>> = OnceCell::new();
+
+    /// Initializes the states of the test [`ENGINE`].
+    ///
+    /// The stacks can be configured with specific stack ops. The stack states are either blank or
+    /// can optionally be updated once like in a freshly started engine.
+    async fn init_engine<'a, I, F>(
+        stack_ops: I,
+        update_after_cleanse: bool,
+    ) -> MappedMutexGuard<'a, Engine>
+    where
+        I: IntoIterator<Item = F> + Send,
+        <I as IntoIterator>::IntoIter: Send,
+        F: FnOnce(&EndpointConfig, &Providers) -> BoxedOps,
+    {
+        let engine = async move {
```

Comment:
:art: 
```suggestion
        let create_engine = async move {
```

it's very easy to end up with the misconception that this will create a engine every time,
e.g. through a `.await?` at the end of the async block.

So renaming it to make clear the the variable contains a lazy constructor of the engine
is a good idea I think.